### PR TITLE
[1781] Higher Level Review show & intake_status Endpoints

### PR DESCRIPTION
**Pull request by camerontesterman**
## Originating Issues
[1781 - Create Higher Level Review Retrieve details endpoint](https://github.com/department-of-veterans-affairs/va.gov-team/issues/1781)
[1777 - Create Higher Level Review Retrieve intake status endpoint](https://github.com/department-of-veterans-affairs/va.gov-team/issues/1771)

## Description of change
Adds two new routes, and methods for handling them, for the appeals controller. The two new routes added are a GET higher level review route, which shows detailed information about a particular higher level review. This leverages the DecisionReview service to call the caseflow API to retrieve data about the specified higher level review. A UUID must be supplied in the request to retrieve the details about the review.

The other route and method added shows the intake status of a particular review. Like the other addition mentioned above, it also utilizes the Decision Review service to make the appropriate call to the caseflow API and likewise requires a UUID supplied in the request path in order to retrieve the status.

## Testing done
Has been tested locally and confirmed using the VCR cassettes that were recorded when the service (DecisionReview) was created.

## Testing planned
None at this time (Dec 17, 2019)

## Acceptance Criteria (Definition of Done)
The two routes work as expected, returning the appropriate data when supplied a valid UUID.

#### Applies to all PRs

- [x] Appropriate logging
- [x] Swagger docs have been updated, if applicable
- [x] Provide link to originating GitHub issue, or connected to it via ZenHub
- [x] Does not contain any sensitive information (i.e. PII/credentials/internal URLs/etc., in logging, hardcoded, or in specs)
- [x] Provide which alerts would indicate a problem with this functionality (if applicable)

### CODE REVIEW

File: spec/request/appeals_request_spec.rb [hunk 1]
```diff
@@ -2,13 +2,75 @@
...
 
+  uuid = '1234567890'
+  ssn = '111223333'
+
+  %w[loa1 loa3].each do |user|
+    define_method("with_#{user}_user") do |options*|
+      using_ssn = options[:without_ssn] ? nil : ssn
+      let(:user) { FactoryBot.create(:user, user.to_sym, ssn: using_ssn) }
+    end
+  end
```
Comment (kfrz) [inline spec/request/appeals_request_spec.rb]: This is hard to understand. It seems you could just define two `shared_contexts` which would read easier.

```
RSpec.shared_context "with_loa1_user" do
  let(:user) { ... }
end
```

Also, what does the `|options*|` splat usage do? I've never seen it used like so and can't find docs. Doesn't seem like you're passing it anywhere

File: app/swagger/requests/appeals.rb [hunk 2]
```diff
@@ -59,6 +59,56 @@ class Appeals
...
+            key :description, 'Bad Gateway: the upstream appeals app returned an invalid response (500+)'
+            schema do
+              key :'$ref', :Errors
+            end
+          end
+        end
+      end
+
+      swagger_path 'v0/appeals/higher_level_review/intake_status/{intake_id}' do
+        operation :get do
```
Comment (kfrz) [inline app/swagger/requests/appeals.rb]: Should this be expanded?

File: app/controllers/v0/appeals_controller.rb [hunk 3]
```diff
@@ -12,5 +12,21 @@ def index
...
+
+    def intake_status
+      intake_status = service.get_higher_level_reviews_intake_status params[:uuid]
+      render json: intake_status
+    end
+
+    private
+
+    def service
+      DecisionReview::Service.new
```
Comment (kfrz) [inline app/controllers/v0/appeals_controller.rb]: Might we define the `DecisionReview` service namespaced under `Appeals` so as to share configuration? It connects to the same backend service (Caseflow) right?

Comment (camerontesterman) [reply app/controllers/v0/appeals_controller.rb]: I think this suggestion makes sense. Wouldn't be a bad idea to raise this with Silvio in his PR, where the service integration work is defined. [PR 3439](https://github.com/department-of-veterans-affairs/vets-api/pull/3439)

**Comment by camerontesterman** [comment]
> Is this still a WIP? If not ready for review please use the draft PR feature.

It isn't ready. Was unaware of that feature, but will use in the future - thanks for letting me know.

File: spec/request/appeals_request_spec.rb (hunk 1 above)
Comment (camerontesterman) [reply spec/request/appeals_request_spec.rb]: I converted this over to use `shared_context`, appreciate the suggestion. The splat wasn't necessary, it has been removed.

File: app/swagger/requests/appeals.rb (hunk 2 above)
Comment (camerontesterman) [reply app/swagger/requests/appeals.rb]: This has been expanded.

File: spec/controllers/v0/higher_level_review_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe HigherLevelReviewController, type: :controller do
+  describe
```
[the remark this answers is not in the document]
Comment (camerontesterman) [reply spec/controllers/v0/higher_level_review_controller_spec.rb]: Methods were added to appeals controller instead of separate higher level review controller, this file has been removed.

File: lib/decision_review/schemas/review.json [hunk 5]
```diff
@@ -193,7 +193,7 @@
                         "enum": ["RequestIssue"]
                       },
                       "id": {
-                        "type": "string"
+                        "type": ["string", "integer"]
```
Comment (SMLuthi) [inline lib/decision_review/schemas/review.json]: Why was this changed to include `integer`s? The response from HLR designates this as only ever being a `string` var.

Comment (camerontesterman) [reply lib/decision_review/schemas/review.json]: Fixed

Comment (johnpaulashenfelter) [reply lib/decision_review/schemas/review.json]: @camerontesterman @SMLuthi I don't have the context of the actual endpoint, but it doesn't look like it _was_ fixed

Comment (camerontesterman) [reply lib/decision_review/schemas/review.json]: @johnpaulashenfelter ok that is weird.. look at this commit https://github.com/department-of-veterans-affairs/vets-api/pull/3426/commits/c8166bdb8c3b0c380bcd6f4b9250747a0f5ee56f

Comment (camerontesterman) [reply lib/decision_review/schemas/review.json]: I did fix it and push the change... I'm not sure why it isn't reflecting that.

Comment (johnpaulashenfelter) [reply lib/decision_review/schemas/review.json]: That is totally weird. I checked this branch locally and it is 100% fine.

File: modules/appeals_api/config/routes.rb [hunk 6]
```diff
@@ -4,7 +4,13 @@
   match '/v0/*path', to: 'application#cors_preflight', via: [:options]
 
   namespace :v0, defaults: { format: 'json' } do
-    resources :appeals, only: [:index]
+    resources :appeals, only: [:index] do
+      collection do
+        get 'higher_level_reviews/:uuid', to: 'appeals#show_higher_level_review'
+        get 'intake_statuses/:intake_id', to: 'appeals#show_intake_status'
+      end
```
Comment (johnpaulashenfelter) [inline modules/appeals_api/config/routes.rb]: I believe it would have been more RESTy to do something similar to

```
    resources :appeals, only: [:index] do
        resources 'higher_level_reviews/:uuid', only: :show
        resources 'intake_statuses/:intake_id',  only: :show
```

but talked to @camerontesterman and they decided against it earlier on. Seems like a big ask to change at this point.

File: modules/appeals_api/app/controllers/appeals_api/v0/appeals_controller.rb [hunk 7]
```diff
@@ -20,6 +20,16 @@ def index
         )
       end
 
+      def show_higher_level_review
+        higher_level_review = review_service.get_higher_level_reviews(params[:uuid])
```
Comment (kfrz) [inline modules/appeals_api/app/controllers/appeals_api/v0/appeals_controller.rb]: This is an anti-pattern in conventional Rails controller. We should consider opening a separate controller that maps more appropriately to this resource - i.e. `/v0/higher_level_review` 

Perhaps y'all have an argument otherwise - but I don't see why this can't be it's own resource

File: modules/appeals_api/config/routes.rb [hunk 8]
```diff
@@ -4,7 +4,13 @@
   match '/v0/*path', to: 'application#cors_preflight', via: [:options]
 
   namespace :v0, defaults: { format: 'json' } do
-    resources :appeals, only: [:index]
+    resources :appeals, only: [:index] do
+      collection do
+        get 'higher_level_reviews/:uuid', to: 'appeals#show_higher_level_review'
```
Comment (kfrz) [inline modules/appeals_api/config/routes.rb]: See comment above about separate resources

**Comment by kfrz** [comment]
:joy_cat: timing ! @SMLuthi could you please update this PR with a comment explaining your rationale for *not* adding a more RESTful resource route? Mostly for historical documentation purposes but also because I may not be seeing a compelling Railsy argument in favor of the current code
